File: akka-docs/rst/java/code/docs/stream/javadsl/cookbook/RecipeDecompress.java
/**
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com/>
 */
package docs.stream.javadsl.cookbook;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Compression;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.testkit.JavaTestKit;
import akka.util.ByteString;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.concurrent.TimeUnit;
import java.util.zip.GZIPOutputStream;

public class RecipeDecompress extends RecipeTest {

    static ActorSystem system;
    static Materializer mat;

    @BeforeClass
    public static void setup() {
        system = ActorSystem.create("RecipeDecompress");
        mat = ActorMaterializer.create(system);
    }

    @AfterClass
    public static void tearDown() {
        JavaTestKit.shutdownActorSystem(system);
        system = null;
        mat = null;
    }

    private ByteString gzip(final String s) throws IOException {
        final ByteArrayOutputStream buf = new ByteArrayOutputStream();
        final GZIPOutputStream out = new GZIPOutputStream(buf);
        try {
            out.write(s.getBytes(StandardCharsets.UTF_8));
        } finally {
            out.close();
        }
        return ByteString.fromArray(buf.toByteArray());
    }

    @Test
    public void parseLines() throws Exception {
        final Source<ByteString, NotUsed> compressed = Source.single(gzip("Hello World"));

        //#decompress-gzip
        final Source<String, NotUsed> uncompressed = compressed
                .via(Compression.gunzip(100))
                .map(b -> b.utf8String());
        //#decompress-gzip

        uncompressed.runWith(Sink.head(), mat).toCompletableFuture().get(1, TimeUnit.SECONDS);
    }

}

File: akka-docs/rst/scala/code/docs/stream/cookbook/RecipeDecompress.scala
/**
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com/>
 */
package docs.stream.cookbook

import java.io.ByteArrayOutputStream
import java.nio.charset.StandardCharsets
import java.util.zip.GZIPOutputStream

import akka.stream.impl.io.compression.GzipCompressor
import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source
import akka.util.ByteString

import scala.annotation.tailrec
import scala.concurrent.Await
import scala.concurrent.duration._

class RecipeDecompress extends RecipeSpec {
  def gzip(s: String): ByteString = {
    val buf = new ByteArrayOutputStream()
    val out = new GZIPOutputStream(buf)
    try out.write(s.getBytes(StandardCharsets.UTF_8)) finally out.close()
    ByteString(buf.toByteArray)
  }

  "Recipe for decompressing a Gzip stream" must {
    "work" in {
      val compressed = Source.single(gzip("Hello World"))

      //#decompress-gzip
      import akka.stream.scaladsl.Compression
      val uncompressed = compressed.via(Compression.gunzip())
        .map(_.utf8String)
      //#decompress-gzip

      Await.result(uncompressed.runWith(Sink.head), 3.seconds) should be("Hello World")
    }
  }
}
